packages/kit/src/core/route_preprocessor.js

```diff
@@ -0,0 +1,55 @@
+import path from 'path';
+import MagicString from 'magic-string';
+import { normalizePath } from 'vite';
+
+/**
+ * Injects `use:__handle_route` from `@sveltejs/kit/router` to handle scrolling when route component mounts
+ * @param {import('types/config').ValidatedConfig} config
+ * @returns {import('svelte/types/compiler/preprocess').PreprocessorGroup}
+ */
+export function route_preprocessor(config) {
+	// Vite normalizes `filename` so we normalize the config's as well
+	const routesPath = normalizePath(config.kit.files.routes);
+
+	return {
+		markup({ content, filename }) {
+			if (
+				filename &&
+				filename.startsWith(routesPath) &&
+				config.extensions.some((ext) => filename.endsWith(ext)) &&
+				!path.basename(filename).startsWith('_')
+			) {
+				const s = new MagicString(content);
+
+				const svelte_body_regex = /<svelte:body[^>]*\/>|<svelte:body.*<\/svelte:body>/;
```

Comment:
do we need to do this in a preprocessor or could we just put it in the root component?

https://github.com/sveltejs/kit/blob/76d26ecaf371c976330263c6da1464316165a05b/packages/kit/src/core/create_app/index.js#L145

Reply:
Interesting, I'm not too familiar with that but it looks like it's only mounted once throughout the app lifetime. The scroll handling would need to happen whenever a new route component mounts. Nonetheless, I'm still having trouble getting this PR done. The `svelte:body` approach isn't as robust as I thought since it would fail if the developer already used it in the markup. I'd need some more time to find alternatives.

Reply:
@bluwy Couldn't the action could return an `update` function that runs on every navigation?

Reply:
We'd need the scrolling to happen exactly between "render new route" and "onMount". I'm not sure if the `update` function guarantees that. I haven't had time to revisit this lately, and my solution for this PR is flawed, so unfortunately it'll be stalled for a while more.